sql: add DROP {CONSTRAINT,COLUMN} support

tamird opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/2929)

<!-- Reviewable:end -->

tamird committed:
DRY

tamird committed:
sql: add DROP CONSTRAINT support

tamird committed:
sql: fix outdated comment

tamird committed:
sql: add DROP [COLUMN] support

vivekmenezes commented:
hey can you wait until my PR is in? It's going to take a day or two.

On Tue, Oct 27, 2015 at 3:14 PM Tamir Duberstein <notifications@github.com>
wrote:

> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/cockroachdb/cockroach/pull/2929
> Commit Summary
>
>    - DRY
>    - sql: fix outdated comment
>    - sql: add DROP CONSTRAINT support
>    - sql: add DROP [COLUMN] support
>
> File Changes
>
>    - *M* sql/alter_table.go
>    <https://github.com/cockroachdb/cockroach/pull/2929/files#diff-0> (92)
>    - *M* sql/backfill.go
>    <https://github.com/cockroachdb/cockroach/pull/2929/files#diff-1> (55)
>    - *M* sql/create.go
>    <https://github.com/cockroachdb/cockroach/pull/2929/files#diff-2> (20)
>    - *M* sql/delete.go
>    <https://github.com/cockroachdb/cockroach/pull/2929/files#diff-3> (4)
>    - *M* sql/drop.go
>    <https://github.com/cockroachdb/cockroach/pull/2929/files#diff-4> (2)
>    - *M* sql/rename.go
>    <https://github.com/cockroachdb/cockroach/pull/2929/files#diff-5> (2)
>    - *M* sql/structured.go
>    <https://github.com/cockroachdb/cockroach/pull/2929/files#diff-6> (4)
>    - *M* sql/testdata/alter_table
>    <https://github.com/cockroachdb/cockroach/pull/2929/files#diff-7> (55)
>
> Patch Links:
>
>    - https://github.com/cockroachdb/cockroach/pull/2929.patch
>    - https://github.com/cockroachdb/cockroach/pull/2929.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2929>.
>

petermattis commented:
We need to do a better job of coordinating work. Someone is going to have an ugly merge. 
<!-- Reviewable comment -K1eo82r3HTYkDT_hOjI:916816231 -->

---

Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[sql/alter_table.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1emcIfXHdfxaOzjiad)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L103)):</sup>
Add a separating blank line before the case statement for consistency.
<!-- Reviewable comment -K1emcIfXHdfxaOzjiae:2108785089 -->

---

<sup>**[sql/alter_table.go, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1enES43iAcJ-0apDW4)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L114)):</sup>
`s/is contained in/is referenced by/g`? Not sure about this.
<!-- Reviewable comment -K1enES43iAcJ-0apDW5:-1836350910 -->

---

<sup>**[sql/alter_table.go, line 131 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1en4vENLuZqvEjOCjQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L131)):</sup>
This can't happen, right? Similar to another comment, perhaps `FindColumnByName` should return the position of the column.
<!-- Reviewable comment -K1en4vENLuZqvEjOCjR:-799085464 -->

---

<sup>**[sql/alter_table.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1empg31FV3-iwSQ9-L)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L153)):</sup>
This can't happen, right? Perhaps `FindIndexByName` could return the position of the index.
<!-- Reviewable comment -K1empg4X5AVqdl0Qr0Y:-246508076 -->

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
I think this is only correct because you know how columns are being removed which feels a bit fragile. I'd prefer passing in the list of dropped columns explicitly, just as you used to do with the add indexes. 
<!-- Reviewable comment -K1engH1q56Mbc-0QatM:-1242568878 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

tamird committed:
sql: FindIndexByName returns an index

tamird commented:
Not it!
<!-- Reviewable comment -K1ez0xP98lStg94mpDI:1645670596 -->

---

Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[sql/alter_table.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1emcIfXHdfxaOzjiad)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L103)):</sup>
Done.
<!-- Reviewable comment -K1ez-x9-YNIRgKEq-Pk:-499146009 -->

---

<sup>**[sql/alter_table.go, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1enES43iAcJ-0apDW4)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L114)):</sup>
Done.
<!-- Reviewable comment -K1etxrJcmtS1U8DNb8g:-499146009 -->

---

<sup>**[sql/alter_table.go, line 131 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1en4vENLuZqvEjOCjQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L131)):</sup>
Done.
<!-- Reviewable comment -K1eyslUqLFhPhfVxQuk:-499146009 -->

---

<sup>**[sql/alter_table.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1empg31FV3-iwSQ9-L)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L153)):</sup>
Done.
<!-- Reviewable comment -K1eys72iVg_9o5QJYJ4:-499146009 -->

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
I think the assumption that columns are not reordered is a valid one, don't you?
<!-- Reviewable comment -K1eu7cpMoV97Jddc3cS:1628724560 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[sql/alter_table.go, line 131 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1en4vENLuZqvEjOCjQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L131)):</sup>
What was done? I'm not seeing any changes here.
<!-- Reviewable comment -K1ifbA68jOAhF4c5eut:-1216753877 -->

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
It's not called out anywhere. And since this is only called from one place, it doesn't seem any easier to regenerate the list of dropped columns vs keeping track of it as they are dropped.
<!-- Reviewable comment -K1ig6BbFdA-k7MnDXJU:79062355 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 9 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
do you want to panic if j >= len(newTableDesc.Columns)? It's probably better to make this code more self contained.
<!-- Reviewable comment -K1inar_ky0mZKwCnF67:-1673151700 -->

---

<sup>**[sql/backfill.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1ioGBIXgmk66MHQN6Q)** ([raw file](https://github.com/cockroachdb/cockroach/blob/a448d8ea3e6d820d3b0d7e3aa5e0588d65f92fa7/sql/backfill.go#L43)):</sup>
Add a comment with some sql examples of what this is doing.
<!-- Reviewable comment -K1ioGBIXgmk66MHQN6R:-445244307 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[sql/backfill.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1ipQhzj8crffP10FZp)** ([raw file](https://github.com/cockroachdb/cockroach/blob/a448d8ea3e6d820d3b0d7e3aa5e0588d65f92fa7/sql/backfill.go#L48)):</sup>
Does this work if you're dropping more than one column? I would think you would need something like `UPDATE <table> SET (a, b) = (NULL, NULL)`
<!-- Reviewable comment -K1ipQi-yhjs1dvlxoWb:-1322743613 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

tamird committed:
sql: FindColumnByName returns an index

tamird commented:
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[sql/alter_table.go, line 131 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1en4vENLuZqvEjOCjQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L131)):</sup>
Oh, this one wasn't done. Done now.
<!-- Reviewable comment -K1is6fba7WabTfZJ3su:520739294 -->

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
Peter: it's called from two places. Vivek: Fixed this to be safe in that case.
<!-- Reviewable comment -K1iufbHiWJ9xLkQfeDf:956547885 -->

---

<sup>**[sql/backfill.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1ioGBIXgmk66MHQN6Q)** ([raw file](https://github.com/cockroachdb/cockroach/blob/a448d8ea3e6d820d3b0d7e3aa5e0588d65f92fa7/sql/backfill.go#L43)):</sup>
Done.
<!-- Reviewable comment -K1j4fE1Xgwr67EnNrxg:-499146009 -->

---

<sup>**[sql/backfill.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1ipQhzj8crffP10FZp)** ([raw file](https://github.com/cockroachdb/cockroach/blob/a448d8ea3e6d820d3b0d7e3aa5e0588d65f92fa7/sql/backfill.go#L48)):</sup>
Fixed and added a test.
<!-- Reviewable comment -K1iwVA5tMPvfSzXqMuG:-1549918397 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
It's called from two places, but only one drops column descriptors.
<!-- Reviewable comment -K1j9d-056vKRi7g5at0:1629672692 -->

---

<sup>**[sql/backfill.go, line 26 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1j9MkIpnRHaKDDHS1D)** ([raw file](https://github.com/cockroachdb/cockroach/blob/78c401c90dba9743168b6aa6bfa7c83d0fc1695c/sql/backfill.go#L26)):</sup>
`s/makecol/makeCol/g`
<!-- Reviewable comment -K1j9MkIpnRHaKDDHS1E:1821540426 -->

---

<sup>**[sql/backfill.go, line 59 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1j9rOH9tdm3QB16w0U)** ([raw file](https://github.com/cockroachdb/cockroach/blob/78c401c90dba9743168b6aa6bfa7c83d0fc1695c/sql/backfill.go#L59)):</sup>
I believe this is actually `SET col1 = NULL, col2 = NULL, ...`
<!-- Reviewable comment -K1j9rOH9tdm3QB16w0V:1763376294 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
Again, that's true, but I really don't want to push that responsibility to the caller. I've added a comment documenting the assumption here.
<!-- Reviewable comment -K1jAwZ3NaxIZjZ-rHCu:1991719022 -->

---

<sup>**[sql/backfill.go, line 26 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1j9MkIpnRHaKDDHS1D)** ([raw file](https://github.com/cockroachdb/cockroach/blob/78c401c90dba9743168b6aa6bfa7c83d0fc1695c/sql/backfill.go#L26)):</sup>
Done.
<!-- Reviewable comment -K1jAutjA5R3lHkpLQd_:-499146009 -->

---

<sup>**[sql/backfill.go, line 59 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1j9rOH9tdm3QB16w0U)** ([raw file](https://github.com/cockroachdb/cockroach/blob/78c401c90dba9743168b6aa6bfa7c83d0fc1695c/sql/backfill.go#L59)):</sup>
That's true, updated.
<!-- Reviewable comment -K1jAM7cXVjTVTixLA9N:1402961661 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

tamird committed:
sql: avoid table scans when no indexes are being added

tamird committed:
sql: move index deletion into `backfillBatch`

vivekmenezes commented on sql/backfill.go:
```diff
@@ -23,28 +23,75 @@ import (
-func (p *planner) makeBackfillBatch(tableName *parser.QualifiedName, tableDesc *TableDescriptor, indexDescs ...IndexDescriptor) (client.Batch, error) {
-	b := client.Batch{}
+func makeColIDtoRowIndex(row planNode, desc *TableDescriptor) (map[ColumnID]int, error) {
+	colIDtoRowIndex := make(map[ColumnID]int)
```
make(, len(row.Columns())) for size of the map

tamird replied:
done

vivekmenezes commented:
LGTM

tamird commented:
Review status: 0 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
I've changed this to sort the descriptors before attempting to work out the ones that were removed.
<!-- Reviewable comment -K1jT-kngTFWGhyBanhx:1127347138 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM
<!-- Reviewable comment -K1jhEPQ-Eqg1kSU1jAn:676834233 -->

---

Review status: 0 of 10 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

<sup>**[sql/backfill.go, line 45 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1jgqlx2sGTe3pkLM_a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/191b6dfbdde9cf39908a1daf36cf642be4c4f7c1/sql/backfill.go#L45)):</sup>
We generally don't put type terminology such as `Slice` in the type name. I would have named this `columnsByID`.
<!-- Reviewable comment -K1jgqlx2sGTe3pkLM_b:-743551780 -->

---

<sup>**[sql/backfill.go, line 57 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1jh0wA6wvH60YOitMj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/191b6dfbdde9cf39908a1daf36cf642be4c4f7c1/sql/backfill.go#L57)):</sup>
`indexesByID`?
<!-- Reviewable comment -K1jh0wA6wvH60YOitMk:-682574826 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 10 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

<sup>**[sql/backfill.go, line 45 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1jgqlx2sGTe3pkLM_a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/191b6dfbdde9cf39908a1daf36cf642be4c4f7c1/sql/backfill.go#L45)):</sup>
Done.
<!-- Reviewable comment -K1jizZowgApaUEiTcGB:-499146009 -->

---

<sup>**[sql/backfill.go, line 57 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1jh0wA6wvH60YOitMj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/191b6dfbdde9cf39908a1daf36cf642be4c4f7c1/sql/backfill.go#L57)):</sup>
Done.
<!-- Reviewable comment -K1jizoDx18hCKfXNX0K:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM, just be sure to coordinate merging with Vivek.
<!-- Reviewable comment -K1jlGfS2DvYBy3yF4zU:809905180 -->

---

Review status: 0 of 10 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)
<!-- Sent from Reviewable.io -->
